src/rust/catnap/mod.rs

```diff
@@ -474,33 +279,22 @@ impl CatnapLibOS {
         trace!("close() qd={:?}", qd);
         let mut qtable: RefMut<IoQueueTable<CatnapQueue>> = self.qtable.borrow_mut();
         match qtable.get_mut(&qd) {
-            Some(queue) => match queue.get_fd() {
-                Some(fd) => {
-                    // Set socket as closing.
-                    let socket: &Socket = queue.get_socket();
-                    let closing_socket: Socket = socket.close()?;
-                    queue.set_socket(&closing_socket);
-
-                    // Close underlying socket.
-                    if unsafe { libc::close(fd) } != 0 {
-                        let errno: libc::c_int = unsafe { *libc::__errno_location() };
-                        let cause: String = format!("failed to close underlying socket (errno={:?})", errno);
-                        error!("close(): {:?}", &cause);
-                        return Err(Fail::new(errno, &cause));
-                    }
-
-                    // Cancel all pending operations.
-                    queue.cancel_pending_ops(Fail::new(libc::ECANCELED, "This queue was closed"));
-
-                    // Update socket state.
-                    let socket: &Socket = queue.get_socket();
-                    let closed_socket: Socket = socket.closed()?;
-                    queue.set_socket(&closed_socket);
-
-                    qtable.free(&qd);
-                    Ok(())
-                },
-                None => unreachable!("CatnapQueue has invalid underlying file descriptor"),
+            Some(queue) => {
+                queue.prepare_close()?;
+                match queue.close() {
+                    Ok(()) => {
+                        queue.commit();
```

Comment:
All of the next 3 lines could probably be rolled up into queue.close()?

Reply:
Yes, they could, but then the responsability for driving the state machine would be spread across the two layers. Eventually we want to place all of them in the queue layer, but this is not possible now.